core/domain/classifier_services.py

```diff
@@ -0,0 +1,534 @@
+                coeff_b = 1.0 / (
+                    self._c_lw.sum(axis=1) + self._num_words * self._beta)
+                label_probabilities = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_label = numpy.random.multinomial(
+                    1,
+                    label_probabilities / label_probabilities.sum()).argmax()
+
+                statez['computes'] += 1
+                if l != new_label:
+                    statez['updates'] += 1
+
+                self._l_dp[d][p] = new_label
+                self._increment_counting_matrices(d, w, new_label)
+
+        return statez
+
+    def _get_label_probabilities(self, d):
+        """
+        Returns a list of label probabilities for a given doc, indexed by
+        label id.
+        """
+        probs_with_label_id = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_with_label_id = (probs_with_label_id /
+            probs_with_label_id.sum(axis=0)[numpy.newaxis])
+        return probs_with_label_id
+
+    def _get_prediction_report_for_doc(self, d):
+        """Generates and returns a prediction report for a given doc.
+
+        The prediction report is a dict with the following keys:
+        - 'prediction_label': the document's predicted label
+        - 'prediction_confidence': the prediction confidence. This is
+        Prob(the doc should be assigned this label |
+        the doc is not assigned _DEFAULT_LABEL).
+        - 'all_predictions': a dict mapping each label to
+        Prob(the doc should be assigned to this label).
+
+        Because _DEFAULT_LABEL is a special label that captures unspecial
+        tokens (how ironic), its probability is not a good predicting
+        indicator. For the current normalization process, it has on
+        average a higher probability than other labels. To combat this,
+        all other labels are normalized as if the default label did not
+        exist. For example, if we have two labels with probability 0.2
+        and 0.3 with the default label having probability 0.5, the
+        normalized probability for the two labels without the default
+        label is 0.4 and 0.6, respectively.
+
+        The prediction threshold is currently defined at the classifier level.
+        A higher prediction threshold indicates that the predictor needs
+        more confidence prior to making a prediction, otherwise it will
+        predict _DEFAULT_LABEL. This will make non-default predictions more
+        accurate, but result in fewer of them.
+        """
+        default_label_id = self._get_label_id(self._DEFAULT_LABEL)
+        prediction_label_id = default_label_id
+        prediction_confidence = 0
+        probs_with_label_id = self._get_label_probabilities(d)
+        default_prob = probs_with_label_id[default_label_id]
```

Comment:
Yeah, though it makes default_prob_comp a bit more confusing to understand. I think normalization coefficient would work, given the doc comments.